src/ray/core_worker/core_worker.cc

```diff
@@ -2594,42 +2599,71 @@ void CoreWorker::HandleRemoveObjectLocationOwner(
-  const auto &callback = [object_id, reply, send_reply_callback](
-                             const absl::flat_hash_set<NodeID> &locations,
-                             int64_t object_size, const std::string &spilled_url,
-                             const NodeID &spilled_node_id, int64_t current_version,
-                             const absl::optional<NodeID> &optional_primary_node_id) {
+
+  const auto callback = [this, object_id](
+                            const absl::flat_hash_set<NodeID> &locations,
+                            int64_t object_size, const std::string &spilled_url,
+                            const NodeID &spilled_node_id, int64_t current_version,
+                            const absl::optional<NodeID> &optional_primary_node_id) {
     auto primary_node_id = optional_primary_node_id.value_or(NodeID::Nil());
     RAY_LOG(DEBUG) << "Replying to HandleGetObjectLocationsOwner for " << object_id
                    << " with location update version " << current_version << ", "
                    << locations.size() << " locations, spilled url: " << spilled_url
                    << ", spilled node ID: " << spilled_node_id
                    << ", and object size: " << object_size
                    << ", and primary node ID: " << primary_node_id;
+    rpc::PubMessage pub_message;
+    pub_message.set_key_id(object_id.Binary());
+    pub_message.set_channel_type(rpc::ChannelType::WORKER_OBJECT_LOCATIONS_CHANNEL);
+    auto object_locations_msg = pub_message.mutable_worker_object_locations_message();
+
     for (const auto &node_id : locations) {
-      reply->add_node_ids(node_id.Binary());
+      object_locations_msg->add_node_ids(node_id.Binary());
     }
-    reply->set_object_size(object_size);
-    reply->set_spilled_url(spilled_url);
-    reply->set_spilled_node_id(spilled_node_id.Binary());
-    reply->set_current_version(current_version);
-    reply->set_primary_node_id(primary_node_id.Binary());
-    send_reply_callback(Status::OK(), nullptr, nullptr);
+    object_locations_msg->set_object_size(object_size);
+    object_locations_msg->set_spilled_url(spilled_url);
+    object_locations_msg->set_spilled_node_id(spilled_node_id.Binary());
+    object_locations_msg->set_current_version(current_version);
+    object_locations_msg->set_primary_node_id(primary_node_id.Binary());
+    object_status_publisher_->Publish(rpc::ChannelType::WORKER_OBJECT_LOCATIONS_CHANNEL,
+                                      pub_message, object_id.Binary());
   };
+
   auto status = reference_counter_->SubscribeObjectLocations(
-      object_id, request.last_version(), callback);
+      object_id, message.last_version(), std::move(callback));
   if (!status.ok()) {
-    send_reply_callback(status, nullptr, nullptr);
+    object_status_publisher_->PublishFailure(
+        rpc::ChannelType::WORKER_OBJECT_LOCATIONS_CHANNEL, object_id.Binary());
+  }
+}
+
+void CoreWorker::HandleGetObjectLocationsOwner(
```

Comment:
This handler is simplified because we don't need the previous structure to be compatible to both subscription + regular lookup.